lib/state_change_syncer.go

```diff
@@ -0,0 +1,726 @@
+
+		stateChangeIndexBuf = append(stateChangeIndexBuf, dbOperationIndexBytes...)
+	}
+
+	// Write the encoded uint32 indexes to the index file.
+	_, err = indexFile.Write(stateChangeIndexBuf)
+	if err != nil {
+		return fmt.Errorf("Error writing to state change index file: %v", err)
+	}
+	if event.IsMempoolFlush {
+		stateChangeSyncer.StateChangeMempoolFileSize += unflushedBytes.StateChangeBytesSize
+	} else {
+		stateChangeSyncer.StateChangeFileSize += unflushedBytes.StateChangeBytesSize
+	}
+
+	// Update unflushed bytes map to remove the flushed bytes.
+	delete(stateChangeSyncer.UnflushedBytes, flushId)
+	return nil
+}
+
+// SyncMempoolToStateSyncer flushes all mempool transactions to the db, capturing those state changes
+// in the mempool state change file. It also loops through all unconnected transactions and their associated
+// utxo ops and adds them to the mempool state change file.
+func (stateChangeSyncer *StateChangeSyncer) SyncMempoolToStateSyncer(server *Server) (bool, error) {
+	originalCommittedFlushId := stateChangeSyncer.BlockSyncFlushId
+
+	if server.mempool.stopped {
+		return true, nil
+	}
+
+	blockHeight := uint64(server.blockchain.bestChain[len(server.blockchain.bestChain)-1].Height)
+
+	stateChangeSyncer.BlockHeight = blockHeight
+
+	mempoolUtxoView, err := server.GetMempool().GetAugmentedUniversalView()
+	if err != nil {
+		return false, errors.Wrapf(err, "StateChangeSyncer.SyncMempoolToStateSyncer: ")
+	}
+	// Kill the snapshot so that it doesn't affect the original snapshot.
+	mempoolUtxoView.Snapshot = nil
+	mempoolUtxoView.IsMempoolView = true
+
+	// A new transaction is created so that we can simulate writes to the db without actually writing to the db.
+	// Using the transaction here rather than a stubbed badger db allows the process to query the db for any entries
+	// inserted during the flush process. This is necessary to get ancestral records for an entry that is being modified
+	// more than once in the mempool transactions.
+	txn := server.blockchain.db.NewTransaction(true)
+	defer txn.Discard()
+
+	err = mempoolUtxoView.FlushToDbWithTxn(txn, uint64(server.blockchain.bestChain[len(server.blockchain.bestChain)-1].Height))
+
+	if err != nil || originalCommittedFlushId != stateChangeSyncer.BlockSyncFlushId {
+		mempoolUtxoView.EventManager.stateSyncerFlushed(&StateSyncerFlushedEvent{
+			FlushId:        uuid.Nil,
+			Succeeded:      false,
+			IsMempoolFlush: true,
+		})
+		return false, errors.Wrapf(err, "StateChangeSyncer.SyncMempoolToStateSyncer: ")
+	}
+	mempoolUtxoView.EventManager.stateSyncerFlushed(&StateSyncerFlushedEvent{
+		FlushId:          uuid.Nil,
+		Succeeded:        true,
+		IsMempoolFlush:   true,
+		BlockSyncFlushId: originalCommittedFlushId,
+	})
+
+	// Loop through all the unconnected transactions in the mempool and connect them and their utxo ops to the mempool view.
+	for txHash, unconnectedTxn := range server.mempool.unconnectedTxns {
```

Comment:
I'm a bit confused about what this is doing. In particular, I don't think we'll ever have txns in unconnectedTxns the way the mempool works right now. This is because, if a txn doesn't connect to the UniversalView, we just reject it. So a txn is either in the UniversalView or it isn't in the mempool at all.

Was there something you ran into that made you need unconnectedTxns?

Reply:
The intention of this code was to iterate over all transactions in the mempool and emit them to the state syncer. This has been switched from `unconnectedTxns` to `_getTransactionsOrderedByTimeAdded`.